src/main/resources/apigee/__init__.py

```diff
@@ -23,39 +25,75 @@ def __init__(self, organization, target_environment):
             print(resp.status_code)
             print(resp.json())
-            raise Exception("Error during deployment of Apigee API Proxy: %s/%s to environment %s" %
-                (api_proxy, api_proxy_revision, self.target_environment))
+            raise Exception("Error during deployment of Apigee API Proxy: %s/%s to environment %s" % (api_proxy, api_proxy_revision, self.target_environment))
         return resp
 
     def undeploy(self, api_proxy, api_proxy_revision):
         revision = self.parse_revision(api_proxy_revision)
         url = self.build_url(api_proxy, revision)
         url = url + "/deployments"
         print(url)
-        resp = requests.delete(url, verify=False, auth=self.authentication)
+        authorization_headers = self.build_authorization_header()
+        headers = authorization_headers
+        if self.mfa:
+            print("Multi factor authentication is on")
+            resp = requests.delete(url, verify=False, headers=headers)
+        else:
+            print("Multi factor authentication is off")
+            resp = requests.delete(url, verify=False, auth=self.authentication)
         if resp.status_code > 399:
             print(resp.status_code)
             print(resp.json())
-            raise Exception("Error during undeployment of Apigee API Proxy: %s/%s from environment %s" %
-                (api_proxy, api_proxy_revision, self.target_environment))
+            raise Exception("Error during undeployment of Apigee API Proxy: %s/%s from environment %s" % (api_proxy, api_proxy_revision, self.target_environment))
         return resp
 
+    def create_one_time_password(self):
+        my_secret = self.organization.secretKey
+        if my_secret is None:
+            raise Exception("Error during creating one time password. The secret key is empty.")
+        my_token = otp.get_totp(my_secret)
+        if len(str(my_token)) == 5:
+            my_token = "0%s" % my_token
+        return my_token
+
+    def build_authorization_header(self):
+        authorization_headers = {}
+        if self.mfa:
+            my_token = self.create_one_time_password()
+            headers = {'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8', 'Accept': 'application/json;charset=utf-8', 'Authorization': 'Basic ZWRnZWNsaTplZGdlY2xpc2VjcmV0'}
+            params = "{'username': '%s', 'password': '%s', 'grant_type': 'password', 'mfa_token': '%s'}" % (self.organization.username, self.organization.password, my_token)
+            resp = requests.post(self.sso_login_url, params=ast.literal_eval(params), verify=False, headers=headers)
+            if resp.status_code > 399:
+                print(resp.status_code)
+                print(resp.json())
+                raise Exception("Error during creating authorization header")
+            data = resp.json()
+            access_token = data['access_token']
+            authorization_headers = "{'Authorization': 'Bearer %s'}" % (access_token)
```

Comment:
Here also, why not use a dict directly?

Reply:
See above why you don't want to do this.

Reply:
Done.